bindings/azure/openai/openai.go

```diff
@@ -0,0 +1,283 @@
+		if err != nil {
+			return nil, fmt.Errorf("error executing %s: %w", &req.Data, err)
+		}
+		p.logger.Infof("response is : %v", response)
+		resp.Metadata["response"] = response
+
+	default:
+		return nil, fmt.Errorf(
+			"invalid operation type: %s. Expected %s, %s, %s",
+			req.Operation, CompletionOperation, ChatCompletionOperation,
+		)
+	}
+
+	endTime := time.Now().UTC()
+	resp.Metadata["end-time"] = endTime.Format(time.RFC3339Nano)
+	resp.Metadata["duration"] = endTime.Sub(startTime).String()
+
+	return resp, nil
+}
+
+func (s *ChatSettings) Decode(in interface{}) error {
+	if err := config.Decode(in, s); err != nil {
+		return fmt.Errorf("decode failed. %w", err)
+	}
+
+	return nil
+}
+
+func (p *AzOpenAI) completion(ctx context.Context, message []byte, metadata map[string]string) (response string, err error) {
+	var ma Prompt
+	err = json.Unmarshal(message, &ma)
+	if err != nil {
+		return "", fmt.Errorf("error in unmarshalling the message array: %w", err)
+	}
+
+	settings := ChatSettings{}
+	settings.Temperature = 1.0
+	settings.TopP = 1.0
+	settings.MaxTokens = 16
+	settings.N = 1
+
+	err = settings.Decode(metadata)
+	if err != nil {
+		return "", fmt.Errorf("Error in decoding the parameters: %w", err)
+	}
+
+	resp, err := p.client.GetCompletions(context.TODO(), azopenai.CompletionsOptions{
+		Prompt:      []*string{to.Ptr(ma.Prompt)},
+		MaxTokens:   to.Ptr(int32(15)),
+		Temperature: to.Ptr(float32(0.0)),
+	}, nil)
+
+	if err != nil {
+		panic(err)
+	}
+	entry := resp.Completions
+	response = *entry.Choices[0].Text
+
+	return
+}
+
+func (p *AzOpenAI) chatCompletion(ctx context.Context, messageArray []byte, metadata map[string]string) (response string, err error) {
+	var ma MessageArray
+	err = json.Unmarshal(messageArray, &ma)
+	if err != nil {
+		return "", fmt.Errorf("error in unmarshalling the message array: %w", err)
+	}
+
+	settings := ChatSettings{}
+	settings.Temperature = 1.0
+	settings.TopP = 1.0
+	settings.N = 1
+
+	err = settings.Decode(metadata)
+	if err != nil {
+		return "", fmt.Errorf("Error in decoding the parameters: %w", err)
+	}
+
+	messageReq := []*azopenai.ChatMessage{}
+	for _, message := range ma.Messages {
+		role := azopenai.ChatRole(message.Role)
+		messageReq = append(messageReq, &azopenai.ChatMessage{
+			Role:    to.Ptr(role),
+			Content: to.Ptr(message.Message),
+		})
+	}
+	var maxTokens *int32
+	if settings.MaxTokens != 0 {
+		maxTokens = to.Ptr(int32(settings.MaxTokens))
+	}
+
+	res, err := p.client.GetChatCompletions(context.TODO(), azopenai.ChatCompletionsOptions{
+		MaxTokens:   maxTokens,
+		Temperature: to.Ptr(float32(settings.Temperature)),
+		TopP:        to.Ptr(float32(settings.TopP)),
+		N:           to.Ptr(int32(settings.N)),
+		Messages:    messageReq,
+	}, nil)
+
+	if err != nil {
+		return "", fmt.Errorf("Error in chat completion api: %w", err)
+	}
+
+	entry := res.ChatCompletions
+	response = *entry.Choices[0].Message.Content
```

Comment:
Please add nil-checks and also checks to make sure that `entry.Choices` has entry 0

Reply:
Done.